Assembly cleanup for ARM64

carlocaione opened the pull request:
Two commits:
- The first one is just a cleanup commit to rework a bit how registers are used. Hopefully now the registers allocation makes a bit more sense.
- The second one is a rework of two macros. This is useful when we do not want to clobber some volatile registers.

carlocaione committed:
arch: arm64: Rework registers allocation

Rationalize the registers usage trying to reuse the smallest set of
registers possible.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

sandeepbrcm commented on arch/arm/core/aarch64/reset.S:
```diff
@@ -54,12 +54,12 @@ GTEXT(__start)
 
 	/* Setup vector table */
-	adr	x19, _vector_table
+	adr	x2, _vector_table
```
'x2' can be clobbered, we need to use a callee saved reg as now we provide a hook for platform to do el3 specific init if needed. ` z_arch_el3_plat_init`

carlocaione replied:
missed the call to `z_arch_el3_plat_init`. ACK. Fix coming.

sandeepbrcm commented on arch/arm/core/aarch64/macro_priv.inc:
```diff
@@ -111,17 +112,56 @@
+ * @brief Save volatile registers for cooperative switching
+ *
+ * This is the equivalent of z_arm64_enter_exc() but used for cooperative switching
+ * in z_arm64_svc().  In this case we do not need to save all the x0-x18
```
This seems right in the current context where SVC is being used to mimic a pendSV and achieving the context switching.
However in a generic SVC handler we need to save-restore all the registers like any other exception right, compiler will not do anything for the SVC instruction itself  ?

carlocaione replied:
The compiler will not do anything for the SVC instruction itself, it's up to us saving the registers. We need to preserve all the registers only for the preemptive context switch. For the cooperative switch it's not needed to preserve all the registers. The SVC call is only used for cooperative context switch or IRQ offloading, in both cases we only need a small subset of the registers saved, in all the other cases we are triggering a fatal error. In fact we currently do not have a generic SVC call handling code in place, we just trigger an error when we do not recognize a valid SVC parameter.

Please also note that this PR is a bit more than a cleanup. This commit is a preliminary work for the switch to the new `arch_switch()` API (that I have ready already) where the `x0` and `x1` registers content passed to `z_arm64_svc()` are being actively used in the thread switch code.

sandeepbrcm replied:
Rest of the changes in the PR are clear to me. +1  Only this optimization of skipping save restore of corruptible registers in SVC handler is bit unconventional.  This skipping save restore is presumably not a necessary step for the next PR you mentioned.  I do agree there are no generic SVC usage as of now. But it can be in future. That time someone will have to add these back again.

And what happens to the irq offload case. The offload handlers may corrupt the x0-x18 ? Or they don't.  Just making sure.
I have not really looked into that path.

wentongwu replied:
I think the reason we don't need save x0-x18 is that there is no use x0-x18 within function z_arm64_call_svc, especially after `svc` instruction there is no use these caller-saved registers.
```
GTEXT(z_arm64_call_svc)
SECTION_FUNC(TEXT, z_arm64_call_svc)
        svc     #_SVC_CALL_CONTEXT_SWITCH
        ret
```

similar reason for z_arm64_offload.

carlocaione replied:
> And what happens to the irq offload case. The offload handlers may corrupt the x0-x18 ? Or they don't. Just making sure.

x0-x18 are caller-saved registers, that means that it is the caller's responsibility to push these registers onto the stack. This is true for the cooperative switch initiated with `arch_swap()` and for the IRQ offloading initiated with `arch_irq_offload()`. 

So, if I'm not wrong, each thread calling into `arch_swap()` or `arch_irq_offload()` already expects the registers x0-x18 to be corrupted eventually, so saving and restoring those is wasted work.

sandeepbrcm replied:
>x0-x18 are caller-saved registers, that means that it is the caller's responsibility to push these registers onto the stack.

Compiler knows function like `z_arm64_call_svc` do not clobber all those registers but that is not true because  `svc `can.
I think weather it will affect something depends on how its used.  In this particular case of 'arch_awap()' (how its coded today) it looks safe and it works for sure.

Since other OS generally do save restore on a svc exception not doing so needs conformance in general. So its up to you to drop this svc handler save restore optimization part or keep it if confident about compiler behavior.

carlocaione replied:
> Compiler knows function like `z_arm64_call_svc` do not clobber all those registers but that is not true because `svc `can.

The SVC handler can clobber the caller-saved registers but that wouldn't matter to the calling function.

@sandeepbrcm anyway, I'll drop the second commit. I'll add a new cleanup commit so we can finally merge this (hopefully) and I can push the other stuff I have ready. Thanks for the review.

carlocaione committed:
arch: arm64: Parametrize registers usage for z_arm64_{enter,exit}_exc

Make explicit what registers we are going to be touched / modified when
using z_arm64_enter_exc and z_arm64_exit_exc.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione commented:
@ioannisg can you merge this?

ioannisg commented:
> @ioannisg can you merge this?

The patch looks fine to me, and I guess that since QEMU runs the tests properly, this sort of verifies the correctness.
However, it would be great if some developer on AARC64 extends the tests/arch/arm/arm_thread_swap, to cover the AARCH64 architecture. Test coverage will provide extra confidence in any additional attempt to optimize the thread context-switch code.

ioannisg commented:
@stephanosio , maybe take a quick look?

ioannisg reviewed (approved):
LGTM

stephanosio reviewed (approved):
LGTM
